flink-ml-api/src/test/java/org/apache/flink/ml/api/core/StageTest.java

```diff
@@ -0,0 +1,345 @@
+                        "Description",
+                        new Long[] {8L, 9L},
+                        ParamValidators.alwaysTrue());
+
+        Param<Float[]> FLOAT_ARRAY_PARAM =
+                new FloatArrayParam("floatArrayParam", "Description", new Float[] {10.0f, 11.0f});
+
+        Param<Double[]> DOUBLE_ARRAY_PARAM =
+                new DoubleArrayParam(
+                        "doubleArrayParam",
+                        "Description",
+                        new Double[] {12.0, 13.0},
+                        ParamValidators.alwaysTrue());
+
+        Param<String[]> STRING_ARRAY_PARAM =
+                new StringArrayParam("stringArrayParam", "Description", new String[] {"14", "15"});
+    }
+
+    // A Stage subclass which inherits all parameters from MyParams and defines an extra parameter.
+    private static class MyStage implements Stage<MyStage>, MyParams<MyStage> {
+        private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+        public final Param<Integer> extraIntParam =
+                new IntParam("extraIntParam", "Description", 20, ParamValidators.alwaysTrue());
+
+        public MyStage() {
+            ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+        }
+
+        @Override
+        public Map<Param<?>, Object> getParamMap() {
+            return paramMap;
+        }
+
+        @Override
+        public void save(String path) throws IOException {
+            ReadWriteUtils.saveMetadata(this, path);
+        }
+
+        public static MyStage load(String path) throws IOException {
+            return ReadWriteUtils.loadStageParam(path);
+        }
+    }
+
+    // A Stage subclass without the static load() method.
+    private static class MyStageWithoutLoad implements Stage<MyStage>, MyParams<MyStage> {
+        private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+        public MyStageWithoutLoad() {
+            ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+        }
+
+        @Override
+        public void save(String path) throws IOException {
+            ReadWriteUtils.saveMetadata(this, path);
+        }
+
+        @Override
+        public Map<Param<?>, Object> getParamMap() {
+            return paramMap;
+        }
+    }
+
+    // Asserts that m1 and m2 are equivalent.
+    private static void assertParamMapEquals(Map<Param<?>, Object> m1, Map<Param<?>, Object> m2) {
+        Assert.assertTrue(m1 != null && m2 != null);
+        Assert.assertEquals(m1.size(), m2.size());
+
+        for (Map.Entry<Param<?>, Object> entry : m1.entrySet()) {
+            Assert.assertTrue(m2.containsKey(entry.getKey()));
+            Object v1 = entry.getValue();
+            Object v2 = m2.get(entry.getKey());
+            if (v1.getClass().isArray() && v2.getClass().isArray()) {
+                Assert.assertArrayEquals((Object[]) v1, (Object[]) v2);
+            } else {
+                Assert.assertEquals(v1, v2);
+            }
+        }
+    }
```

Comment:
`HashMap.equals()` does not work here. If the value of an entry is List<...>, then HashMap.equals() will compare the reference of these two arrays instead of comparing the values of these two arrays. Instead, the existing approach explicitly uses `Assert.assertArrayEquals` to compare the values of these two arrays in this case.

I have verified this behavior by changing the code here to use `HashMap.equals()`.